src/kumo-api.ts

```diff
@@ -189,27 +186,43 @@ export class KumoApi {
           label: zoneTable[serial].label,
           zoneTable: zoneTable[serial],
         };
-        this.log.info('Found device. Serial: %s. Label: %s', device.serial, device.label);
-        this.devices.push(device);
+        let existingDeviceIndex: string|number|undefined = undefined;
+        for (const anExistingDeviceIndex in this.devices) {
+          if (this.devices[anExistingDeviceIndex].serial === device.serial) {
+            existingDeviceIndex = anExistingDeviceIndex;
+            break;
+          }
+        }
+
+        if (existingDeviceIndex !== undefined) {
+          this.log.info('Updated existing device. Serial: %s. Label: %s', device.serial, device.label);
+          this.devices[existingDeviceIndex] = device;
+        } else {
+          this.log.info('Found device. Serial: %s. Label: %s', device.serial, device.label);
+          this.devices.push(device);
+          newDevicesCount += 1;
+        }
       }
 
       if(Object.prototype.hasOwnProperty.call(child, 'children')){
-        this.parseChildren(child.children);
+        newDevicesCount += this.parseChildren(child.children);
       }
     }
+
+    return newDevicesCount;
   }
 
   // Refresh the security token.
-  private async checkSecurityToken(): Promise<boolean> {
+  private async checkSecurityToken(forceRefresh: boolean = false): Promise<boolean> {
     const now = Date.now();
 
-    // If we don't have a security token yet, acquire one before proceeding.
-    if(!this.securityToken && !(await this.acquireSecurityToken())) {
-      return false;
+    // If we don't have a security token yet, acquire one directly.
+    if(!this.securityToken) {
+      return await this.acquireSecurityToken();
     }
 
     // Is it time to refresh? If not, we're good for now.
-    if((now - this.securityTokenTimestamp) < KumoTokenExpirationWindow) {
+    if(!forceRefresh && (now - this.securityTokenTimestamp) < KumoTokenExpirationWindow) {
```

Comment:
Allow forcing checkSecurityToken to pull a new token (and refresh all data) prior to the token expiring, as long as it still passes other throttling guards